auth/src/main/java/com/firebase/ui/auth/ui/email/RegisterEmailActivity.java

```diff
@@ -111,28 +144,141 @@ private void setUpTermsOfService() {
         agreementText.setOnClickListener(new View.OnClickListener() {
             @Override
             public void onClick(View view) {
-                Intent intent = new Intent(Intent.ACTION_VIEW).setData(Uri.parse
-                        (mActivityHelper.getFlowParams().termsOfServiceUrl));
-                startActivity(intent);
+                startActivity(new Intent(Intent.ACTION_VIEW).setData(
+                        Uri.parse(mActivityHelper.getFlowParams().termsOfServiceUrl)));
             }
         });
     }
 
+    private void showEmailAutoCompleteHint() {
+        PendingIntent hintIntent = FirebaseAuthWrapperFactory
+                .getFirebaseAuthWrapper(mActivityHelper.getAppName()).getEmailHintIntent(this);
+        if (hintIntent != null) {
+            try {
+                startIntentSenderForResult(hintIntent.getIntentSender(), RC_HINT, null, 0, 0, 0);
+            } catch (IntentSender.SendIntentException e) {
+                Log.e(TAG, "Unable to start hint intent", e);
+            }
+        }
+    }
+
+    @Override
+    public void onActivityResult(int requestCode, int resultCode, Intent data) {
+        super.onActivityResult(requestCode, resultCode, data);
+        switch (requestCode) {
+            case RC_HINT:
+                if (data != null) {
+                    Credential credential = data.getParcelableExtra(Credential.EXTRA_KEY);
+                    if (credential != null) {
+                        mEmailEditText.setText(credential.getId());
+                        String name = credential.getName();
+                        mNameEditText.setText(name);
+                        if (TextUtils.isEmpty(name)) {
+                            mNameEditText.requestFocus();
+                        } else {
+                            mPasswordEditText.requestFocus();
+                        }
+                    }
+                }
+                break;
+            case RC_SIGN_IN:
+            case RC_WELCOME_BACK_IDP:
+                finish(resultCode, data);
+                break;
+        }
+    }
+
+    @Override
+    public void onFocusChange(View view, boolean hasFocus) {
+        if (hasFocus) return;
```

Comment:
Just a general style comment (first noticed it here) but let's add some more comments in meaty methods like this.  For instance in this method the first line could use a comment like `// Only consider fields losing focus`.